source/kernel/src/services/i2c.rs

```diff
@@ -0,0 +1,685 @@
+        Ok(())
+    }
+}
+
+// === impl Transaction ===
+
+impl Transaction {
+    /// Constructs a new `Transaction` from the provided [`StartTransaction`]
+    /// message, returning the `Transaction` and a [`KConsumer`] for receiving
+    /// [`Transfer`]s within that `Transaction`.
+    ///
+    /// This is intended to be used by server implementations of the
+    /// [`I2cService`] when handling [`StartTransaction`] messages.
+    pub async fn new(
+        StartTransaction { addr, capacity }: StartTransaction,
+    ) -> (Self, KConsumer<Transfer>) {
+        let (tx, rx) = KChannel::new_async(capacity).await.split();
+        let rsp_rx = Reusable::new_async().await;
+        let txn = Transaction {
+            addr,
+            rsp_rx,
+            tx,
+            ended: false,
+        };
+        (txn, rx)
+    }
+
+    /// Read `len` bytes from the I²C bus into `buf`.
+    ///
+    /// Note that, rather than always filling the entire buffer, this method
+    /// takes a `len` argument which specifies the number of bytes to read. This
+    /// is intended to allow callers to reuse the same [`FixedVec`] for multiple
+    /// `read` and [`write`](Self::write) operations.
+    ///
+    /// # Arguments
+    ///
+    /// - `buf`: a [`FixedVec`] buffer into which bytes read from the
+    ///   I²C bus will be written
+    /// - `len`: the number of bytes to read. This must be less than or equal to
+    ///   `buf.len()`.
+    /// - `end`: whether or not to end the transaction. If this is `true`, a
+    ///   `STOP` condition will be sent on the bus once `len` bytes have been
+    ///   read. If this is `false`, a repeated `START` condition will be sent on
+    ///   the bus once `len` bytes have been read.
+    ///
+    /// # Errors
+    ///
+    /// - If an error occurs performing the I²C bus transaction.
+    /// - If `len` is greater than [`buf.capacity()`] - [`buf.len()`].
+    /// - If there is no [`I2cService`] running.
+    ///
+    /// # Cancelation Safety
+    ///
+    /// If this future is dropped, the underlying I²C bus read
+    /// operation may still be performed.
+    ///
+    /// [`FixedVec`]: mnemos_alloc::containers::FixedVec
+    /// [`buf.len()`]: mnemos_alloc::containers::FixedVec::len
+    /// [`buf.capacity()`]: mnemos_alloc::containers::FixedVec::capacity
+    pub async fn read(
+        &mut self,
+        buf: FixedVec<u8>,
+        len: usize,
+        end: bool,
+    ) -> Result<FixedVec<u8>, I2cError> {
+        // if there's already data in the buffer, the available capacity is the
+        // total capacity minus the length of the existing data.
+        let cap = buf.capacity() - buf.len();
+        if cap < len {
+            return Err(self.buf_too_small(OpKind::Read, len, cap));
+        }
+        self.xfer(buf, len, end, OpKind::Read).await
+    }
+
+    /// Write `len` bytes `buf` to the I²C.
+    ///
+    /// Note that, rather than always writing all the bytes in the buffer, this
```

Comment:
I'm less certain we need this extra len for write, is there a case you wouldn't want to "send all" for write?

For read it makes sense: there is no len, only capacity. For write, can't we just use len?

Reply:
hmm, that's a good point, we probably could remove this. good call!